### DISCUSSION TITLE
Server pausing or hibernation

### DISCUSSION THREAD

**Discussion by MercifulCamel** [mbround18/valheim-docker#209] [Ideas]
Hi not sure if this has already been asked for or is already implemented.

But I was wondering if it's possible to make the server paus when there are no users online.
Basically make it check for users and if there are none the server pauses or "freezes" until someone connects.

Loving the work so far

**Comment by mbround18** [comment]
I don't think there is any technical way this could be implemented 0.o Valheim does not support pausing and I think if I set up a udp listener the server would take too long to start back up for the user to connect securely. Maybe this could be solved with a Valheim mod?

**Reply by MercifulCamel** [reply]
Maybe, or if it's possible to make the docker image go into hibernation until someone tries to connect.
I don't really know how it would be possible to implement I only know that itzg/minecraft-server docker image has it so thought it would be possible.

**Reply by mbround18** [reply]
unfortunately that's because Java supports hibernation while c-sharp application built on unity do not. :/ If they did the developer would have to put in special code for that; so that might be a feature request that you might want to toss to the valheim developers.


however if someone else has a potential idea on how to write a proxy-based UDP connection to starting the velheim server please let me know pull request.

**Reply by joryirving** [reply]
I came in here to ask the same thing, due to the same minecraft docker image. It just pauses the Java process until it gets a knock. I was doubly interested, because my valheim container likes to idle at 40% cpu when no one is connected. It looks like this is a feature people are looking for in the actual dedicated server software, but at best, it just unloads chunks and pauses any progression while no players are active. Great for in-game, terrible for my power bill.

**Reply by mbround18** [reply]
This is something you would have to ask from the Valheim developers. Java supports process pausing while C# that is something that would have to be coded into the actual server and not something I could accomplish. 

https://psycodedeveloper.wordpress.com/2019/07/31/how-to-pause-or-resume-a-process-with-c/


While I understand it is a desirable feature and I would love to be able to support it; at this time I cannot support it as I personally either lack the knowledge or insight to pause a C# server or it needs to be coded by the developers.

**Reply by Android3000** [reply]
This is not _quite_ the same thing here, but just to add in case it inspires any similar ideas. I'm running a local dedicated server and via some task scheduler/powershell automation, my PC currently enters sleep mode if there are no players connected for a window of time (~15 mins or so). I take a backup before entering sleep, but I actually don't stop the container or anything... I just throw the PC to into sleep mode, and upon resuming everything has worked fine. No world corruption, no issues - but that's probably just luck so far lol. I probably could and should actually stop/start it, but so far I've found I simply just don't need to.

**Reply by joryirving** [reply]
I'd be interested to know how you do it, but I can't stop the PC, since it's running other containers, doing other things.

**Reply by Android3000** [reply]
Sorry I missed your follow up @LilDrunkenSmurf ! So basically this uses the Valve A2S_INFO server query: https://developer.valvesoftware.com/wiki/Server_queries

I found a snippet on Reddit someone had written in Python to do the basic query. I was just testing at that point and stupidly never copied the link to give proper credit, so thank you unknown Redditor out there. The rest of the automation is in PowerShell. It is by no means resilient code, but it's worked for me so far for my own internal private use. Use at your own peril 😬 

(The `$playerCount = python -c` line is probably the most interesting for you)

```powershell
Add-Type -AssemblyName System.Windows.Forms;

function Log-Formatted($message) {
    "[$(Get-Date -Format "MM/dd/yyyy hh:mm:ss tt")] $message" | Out-File -Append -FilePath C:\path\to\log.txt -Encoding utf8
}

function Check-Players-Active { 
    $playerCount = python -c "import socket; addr = ('127.0.0.1', 2457); message = '\xff\xff\x54\x53\x6f\x75\x72\x63\x65\x20\x45\x6e\x67\x69\x6e\x65\x20\x51\x75\x65\x72\x79\x00'; client_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM); client_socket.settimeout(1.0); client_socket.sendto(message.encode(), addr); data = client_socket.recvfrom(2048)[0].decode('utf-8', errors='ignore')[2:].split(chr(0x0)); print(ord(data[6][0]) if data[6] else 0);"
    if ($LASTEXITCODE -ne 0) {
        Log-Formatted "Received a non-zero exit code. Exiting idle check."        
        exit
    } else {
        Log-Formatted "Got players: $playerCount"
        if ($playerCount -ne 0) {
            Log-Formatted "Server active, exiting check."
            exit
        }    
    }
}

Check-Players-Active
Start-Sleep -Seconds 300
Check-Players-Active
Start-Sleep -Seconds 300
Check-Players-Active
Log-Formatted "Server appears to be idle; backing up and sleeping."
$fileName = "$(Get-Date -Format "yyyyMMdd-HHmmss")-sleep-backup.tar.gz"
docker exec -it valheim-server odin backup /home/steam/.config/unity3d/IronGate/Valheim /home/steam/backups/$($fileName)
Start-Sleep -Seconds 15
Log-Formatted "Backup complete, going to sleep after final player check."
Check-Players-Active
Log-Formatted "Sleeping."
[System.Windows.Forms.Application]::SetSuspendState([System.Windows.Forms.PowerState]::Suspend, $false, $false)
```

**Reply by mbround18** [reply]
> Sorry I missed your follow up @LilDrunkenSmurf ! So basically this uses the Valve A2S_INFO server query: https://developer.valvesoftware.com/wiki/Server_queries
> 
> I found a snippet on Reddit someone had written in Python to do the basic query. I was just testing at that point and stupidly never copied the link to give proper credit, so thank you unknown Redditor out there. The rest of the automation is in PowerShell. It is by no means resilient code, but it's worked for me so far for my own internal private use. Use at your own peril 😬 


This would sleep the server but how would you reinitialize it once a player tries to connect?

**Reply by Android3000** [reply]
In my case, as this is a private family server, I have a link that issues a remote Wake-On-LAN broadcast that wakes the PC up (takes just a few seconds to start back up). Unfortunately nothing built in to automatically issue that upon attempting to connect (as the server wouldn't even be showing up in the server browser anyways at that point). It also has some scheduled tasks that wake it for daily update checks and stuff as well.

The idea above is that folks may be able to use the AS2_INFO server query to build in their own automation that fits their use case (e.g. spinning up/down EC2 instances or whatnot).

**Reply by dghodgson** [reply]
Just wanted to chime in on this.

From what I understand, the server already pauses game state if no players are online, so if you're concerned about the world continuing to progress without you, that shouldn't be an issue.

I confirmed this myself when my buddy logged off while a troll was attacking one of our forward bases. When I logged in 10 minutes later, the Troll was mid-swing and I was right in front of it.

If that's not enough for you (e.g. you want to put your physical server to sleep) then you might need to look in to the A2S_INFO query as above.
